### PR TITLE
i.MX93: enable lpi2c and lpspi

### PR DESCRIPTION
**Pull request by CkovMk**
- Imporve the way we handle mmu region for i.MX93
- Enable LPI2C and LPSPI on i.MX93
- Add rumtime mmio capability to LPI2C driver (to demostrate the concept, see below discussion)

Need discussion:
Should we :
  1. add runtime mmio support to all these drivers (like I did to lpi2c driver in a separate commit), 
or
  2. manually add memroy space in "mmu_regions" array (the old way, but get improved in this pr)?

### CODE REVIEW

**Comment by CkovMk** [comment]
also @JiafeiPan here.

File: boards/arm64/mimx93_evk/mimx93_evk-pinctrl.dtsi
Comment (danieldegrasse) [inline boards/arm64/mimx93_evk/mimx93_evk-pinctrl.dtsi:1]: Please split the board level changes into another commit, after the commit with SOC level changes to enable LPI2C and LPSPI. Ideally, the changed to enable LPUART and LPSPI would also be split, so you would have commits like so:
- soc level changes to enable LPI2C
- board level changes to enable LPI2C
- soc level changes to enable LPSPI
- board level changes to enable LPSPI

Comment (CkovMk) [reply boards/arm64/mimx93_evk/mimx93_evk-pinctrl.dtsi:1]: Sure, will do that later.
btw, what's your opinion about memory mapping?

**Comment by mmahadevan108** [comment]
@CkovMk , I see changed to the lpi2c driver. However no changes were made to the lpspi driver. Any reason this driver was not changed?

**Comment by CkovMk** [comment]
> @CkovMk , I see changes to the lpi2c driver. However no changes were made to the lpspi driver. Any reason this driver was not changed?

@mmahadevan108 No. I changed lpi2c driver to show the idea, because I'm in doubt whether it will be accepted. If we decided to update most of our driver to use mmio, I will update lpspi driver as well. Otherwize, I will simply drop that commit and follow the old approach.

At first, I decided to use mmio because we want to avoid editing `mmu_regions[]` everytime adding or removing an instance (e.g. lpi2c8 on rpi header) in the application. However, since the `mmu_regions[]` array can use information from devicetree and map all `status="okay"` nodes that are compatiable to a driver, the initial problem has gone. So now it is only a matter of preference: do we update drivers to adopt mmio api.

**Review by carlocaione — CHANGES_REQUESTED** [review]
This is not how we should do this. The `device_mmio` API should be used instead without adding a new DT macro to create MMIO regions. If the `device_mmio` for some reason is falling short, we should fix that instead of creating one more layer.

File: soc/arm64/nxp_imx/mimx9/mmu_regions.c [hunk 1]
```diff
@@ -8,6 +8,30 @@
...
+ *        node_id##_P_reg_FOREACH_PROP_ELEM* definitions,
+ *        we can't automate dts node with multiple reg
+ *        entries.
+ *
+ */
+#define MIMX9_DT_MMU_REGION_FLAT_ENTRY(node_id, attrs)  \
+	MMU_REGION_FLAT_ENTRY(DT_NODE_FULL_NAME(node_id), \
+			      DT_REG_ADDR(node_id), \
+			      DT_REG_SIZE(node_id), \
+			      attrs),
```
[the remark this answers is not in the document]
Comment (CkovMk) [reply soc/arm64/nxp_imx/mimx9/mmu_regions.c:32]: This macro is used to generage an entry in `mmu_regions[]` down below. Take this for example:

```c
DT_FOREACH_STATUS_OKAY_VARGS(nxp_kinetis_lpuart,
			  MIMX9_DT_MMU_REGION_FLAT_ENTRY,
			 (MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS))
```

It will generate mmap entries for all LPUART instances with `status = "okay"` set automatically.

In the past we need to create LPUART1 ... LPUART8 manually, and if one of them are disabled in DT, the mmap entry is not removed. This results in a huge array of unused regions and occupy space in mmu translate table.

The `DT_FOREACH_STATUS_OKAY_VARGS` macro only takes token withthis format `(node_id, VA_ARGS)` , so make the `DT_FOREACH` macro happy.

Our most (if not all) drivers has not been updated to use `runtime_device_mmio` APIs, thus these regions are needed to map them before drivers actually access the region. This is common to all our MPU products (i.MX8M, i.MX93, etc.) I will assume these are introduced before `runtime_device_mmio` APIs.

If we update drivers to map their own registers during initliztion, (LPI2C driver as an example in this pr), then this is no longer needed.

[the remark this answers is not in the document]
Comment (CkovMk) [reply soc/arm64/nxp_imx/mimx9/mmu_regions.c:32]: I understand your point. But the transition to `device_mmio` APIs takes time. Adding this macro does not create a new way of mapping regions, but just enhances the `mmu_regions.c` so that existing drivers can be easily enabled on MPU, without need to hack the driver or adding a pile of (maybe unused) regions into that array.

**Comment by CkovMk** [comment]
@danieldegrasse patches are split to: 
- enable runtime mmio
- add driver in soc
- add driver to board
for both lpi2c and lpspi.

@carlocaione marco "MMU_REGION_DT_FLAT_ENTRY" is added to arch/arm and arch/arm64.

File: include/zephyr/arch/arm64/arm_mmu.h [hunk 2]
```diff
@@ -207,6 +207,28 @@ struct arm_mmu_ptables {
...
+ * @code{.c}
+ *      DT_FOREACH_STATUS_OKAY_VARGS(nxp_imx_gpio,
+ *				  MMU_REGION_DT_FLAT_ENTRY,
+ *				 (MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS))
+ * @endcode
+ *
+ * @note  Since devicetree_generated.h does not include
+ *        node_id##_P_reg_FOREACH_PROP_ELEM* definitions,
+ *        we can't automate dts node with multiple reg
+ *        entries.
```
Comment (carlocaione) [inline include/zephyr/arch/arm64/arm_mmu.h:224]: I think this can be achieved using `DT_PROP_HAS_IDX` and `LISTIFY`.

File: soc/arm64/nxp_imx/mimx9/mmu_regions.c [hunk 3]
```diff
@@ -30,21 +30,15 @@ static const struct arm_mmu_region mmu_regions[] = {
...
-			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS),
-
 	MMU_REGION_FLAT_ENTRY("IOMUXC",
 			      DT_REG_ADDR(DT_NODELABEL(iomuxc)),
 			      DT_REG_SIZE(DT_NODELABEL(iomuxc)),
 			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS),
 
+	DT_FOREACH_STATUS_OKAY_VARGS(nxp_kinetis_lpuart,
+			  MMU_REGION_DT_FLAT_ENTRY,
+			 (MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS))
```
Comment (carlocaione) [inline soc/arm64/nxp_imx/mimx9/mmu_regions.c:40]: `MMU_REGION_DT_FLAT_ENTRY`

File: soc/arm64/nxp_imx/mimx8m/mmu_regions.c [hunk 4]
```diff
@@ -25,25 +25,19 @@ static const struct arm_mmu_region mmu_regions[] = {
...
-			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS),
-
 	MMU_REGION_FLAT_ENTRY("IOMUXC",
 			      DT_REG_ADDR(DT_NODELABEL(iomuxc)),
 			      DT_REG_SIZE(DT_NODELABEL(iomuxc)),
 			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS),
+
+	DT_FOREACH_STATUS_OKAY_VARGS(nxp_imx_iuart,
+				  MMU_REGION_DT_FLAT_ENTRY,
+				  (MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS))
```
Comment (carlocaione) [inline soc/arm64/nxp_imx/mimx8m/mmu_regions.c:40]: `MMU_REGION_DT_FLAT_ENTRY`

File: drivers/i2c/i2c_mcux_lpi2c.c [hunk 5]
```diff
@@ -47,6 +58,9 @@ struct mcux_lpi2c_config {
 };
 
 struct mcux_lpi2c_data {
+#ifdef CONFIG_MMU
```
Comment (carlocaione) [inline drivers/i2c/i2c_mcux_lpi2c.c:61]: is this (and the others) `ifdef` really useful? If `!DEVICE_MMIO_IS_IN_RAM` a lot (all?) of defines are already collapsing to NOP

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: soc/arm64/nxp_imx/mimx9/mmu_regions.c (hunk 3 above)
Comment (CkovMk) [reply soc/arm64/nxp_imx/mimx9/mmu_regions.c:40]: `MMU_REGION_DT_FLAT_ENTRY` is meant to be used within `DT_FOREACH_STATUS_OKAY_VARGS`.

Comment (carlocaione) [reply soc/arm64/nxp_imx/mimx9/mmu_regions.c:40]: mmmh, right, can it be the opposite? Adding a `MMU_REGION_DT_FLAT_ENTRY` encapsulating the `DT_FOREACH_STATUS_OKAY_VARGS` ? That would be symmetrical to the current `MMU_REGION_FLAT_ENTRY` macro we have.

File: include/zephyr/arch/arm64/arm_mmu.h (hunk 2 above)
Comment (CkovMk) [reply include/zephyr/arch/arm64/arm_mmu.h:224]: Oh, didn't realize `LISTIFY` was a thing.

```c
LISTIFY(DT_NUM_REGS(node_id), MMU_REGION_DT_FLAT_ENTRY_BY_IDX, (,), node_id, attrs)
```

should be ok. But there's still a limitation: all regs must have same `attrs`. For GIC, this is true. But for external memory controller such as FlexSPI, they usually require different `attrs`. I think maybe the use case is too limited. 

Given that:
- Most peripheral with multiple instances only requires 1 entry per device.
- Devices such as GIC only have 1 instance per SoC and is always enabled, it's okay to write 2 entries for it.
- Memory controller such as FlexSPI are not supported on MPU yet. They could manage mappings themself with `device_mmio` APIs by the time they are enabled.

I think we can leave it as-is. The purpose of this macro is to conditionally map simple peripherals with lots of instances, making the use of existing MCU-ready driver easier to use on MPU.

File: soc/arm64/nxp_imx/mimx9/mmu_regions.c (hunk 3 above)
Comment (CkovMk) [reply soc/arm64/nxp_imx/mimx9/mmu_regions.c:40]: It can be wrapped with another layer of macro. It will looks something like 

```c
MMU_REGION_DT_FLAT_ENTRY_STATUS_OKAY(compatiable, attr)
```

It looks really weird, AFAIK there aren't many macros that takes a `compatiable string` as a parameter.

ps: Please give me a better name lol

File: drivers/i2c/i2c_mcux_lpi2c.c (hunk 5 above)
Comment (CkovMk) [reply drivers/i2c/i2c_mcux_lpi2c.c:61]: Not really understanding the point. What do you suggest to distinguish MPU and MCU?

**Comment by CkovMk** [comment]
CI shows test failures unrelated to this pr.
There are also some format issues (use tabs for indent, etc.) will fix in next update.

File: include/zephyr/arch/arm64/arm_mmu.h (hunk 2 above)
Comment (carlocaione) [reply include/zephyr/arch/arm64/arm_mmu.h:224]: mmh, alright, let's leave this as is for now.

File: soc/arm64/nxp_imx/mimx9/mmu_regions.c (hunk 3 above)
Comment (carlocaione) [reply soc/arm64/nxp_imx/mimx9/mmu_regions.c:40]: Nah, using a compatible here looks fine to me. But I would call it `MMU_REGION_DT_FLAT_ENTRY(compatible, attr)` and rename the current macro to something internal.

File: drivers/i2c/i2c_mcux_lpi2c.c (hunk 5 above)
Comment (carlocaione) [reply drivers/i2c/i2c_mcux_lpi2c.c:61]: What I mean is `DEVICE_MMIO_NAMED_RAM` is already collapsing to nothing when `!defined(CONFIG_MMU)` (see in `include/zephyr/sys/device_mmio.h`) so this `ifdef` is redundant.

Comment (CkovMk) [reply drivers/i2c/i2c_mcux_lpi2c.c:61]: Oh, so just remove it here.

File: soc/arm64/nxp_imx/mimx9/mmu_regions.c (hunk 3 above)
Comment (CkovMk) [reply soc/arm64/nxp_imx/mimx9/mmu_regions.c:40]: most macro with "SUBSYS"_DT_"DO_SOMETHING" macro takes node_id as first parameter. I'm unwilling to break the convention here...

Found this:

```c
#define DT_COMPAT_GET_ANY_STATUS_OKAY(compat)
```

so let's call it `MMU_REGION_DT_COMPAT_FLAT_ENTRY(compat, attr)` or `MMU_REGION_DT_COMPAT_FOREACH_STATUS_OKAY_FLAT_ENTRY(compat, attr)` (...it best describes the purpose, but too loooooong).

**Review by carlocaione — DISMISSED** [review]
LGTM for the `MMU_REGION_` macro changes

**Comment by CkovMk** [comment]
rebase on main to adopt changes to lpspi driver.

**Review by carlocaione — DISMISSED** [review]
LGTM for the MMU_REGION_ macro changes. You maybe want to consider pushing the MMU changes as separate PR.

**Comment by CkovMk** [comment]
> LGTM for the MMU_REGION_ macro changes. You maybe want to consider pushing the MMU changes as separate PR.

split as #66869.

MMU changes have been removed from this pr.

**Comment by CkovMk** [comment]
also remove the MCUX_HAS...

**Review by danieldegrasse — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]
LGTM but not exactly sure why I am the assignee here
